File: SDK_v1.1_RFC003/windi-sdk-v1/emitters/python/windi_emitter.py
# WINDI SDK - Reference Emitter (Python)
# RFC-002: WINDI-SEC-JSON-v1 (HMAC default)
# "Flow is Truth. Content is Sovereign."

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import os
import threading
import time
from dataclasses import dataclass
from typing import Any, Dict, Optional, Literal

Alg = Literal["HMAC-SHA256"]
Shelf = Literal["S1", "S2", "S3", "S4", "S5", "S6", "S7"]

EVENTS = {
    "DOC_CREATED",
    "APPROVAL_REQUESTED",
    "APPROVED",
    "REJECTED",
    "APPROVAL_OVERRIDDEN",
    "DEADLINE_EXCEEDED",
    "DEPENDENCY_LINKED",
    "DEPENDENCY_BLOCKING",
    "STATE_TRANSITION",
}

def _b64(b: bytes) -> str:
    return base64.b64encode(b).decode("ascii")

def _sha256(b: bytes) -> bytes:
    return hashlib.sha256(b).digest()

def _canonical_json(obj: Any) -> bytes:
    return json.dumps(obj, ensure_ascii=False, separators=(",", ":"), sort_keys=True).encode("utf-8")

def _hmac_sha256(key: bytes, msg: bytes) -> bytes:
    return hmac.new(key, msg, hashlib.sha256).digest()

@dataclass(frozen=True)
class WindiEmitterConfig:
    client_id: str
    key_id: str
    csalt_b64: str          # 256-bit recommended (32 bytes) base64
    hmac_key_b64: str       # 256-bit recommended base64
    protocol_version: str = "1.0"
    alg: Alg = "HMAC-SHA256"

class Sequence:
    def __init__(self, start: int = 0):
        self._lock = threading.Lock()
        self._seq = start

    def next(self) -> int:
        with self._lock:
            self._seq += 1
            return self._seq

class WindiEmitter:
    def __init__(self, cfg: WindiEmitterConfig):
        self.cfg = cfg
        self.csalt = base64.b64decode(cfg.csalt_b64)
        self.hmac_key = base64.b64decode(cfg.hmac_key_b64)
        self.seq = Sequence()
        self.cid = _b64(_sha256(cfg.client_id.encode("utf-8")))

    def derive_domain_hash(self, domain_id: str) -> str:
        raw = b"WINDI:DOMAIN:v1" + self.csalt + domain_id.encode("utf-8")
        return _b64(_sha256(raw))

    def derive_doc_fingerprint(self, doc_vector_bytes: bytes) -> str:
        raw = b"WINDI:DOC:v1" + self.csalt + doc_vector_bytes
        return _b64(_sha256(raw))

    def sign_packet(self, header: Dict[str, Any], payload: Dict[str, Any]) -> str:
        signing_obj = {"header": header, "payload": payload}
        mac = _hmac_sha256(self.hmac_key, _canonical_json(signing_obj))
        return _b64(mac)

    def emit(
        self,
        shelf: Shelf,
        code: str,
        weight: int,
        domain_id: str,
        doc_vector_bytes: bytes,
        event: str,
        ctx_window: Optional[str] = "5m",
        ctx_flags: int = 0,
        ts_ms: Optional[int] = None,
    ) -> Dict[str, Any]:
        if event not in EVENTS:
            raise ValueError(f"Invalid event: {event}")
        if not (0 <= weight <= 100):
            raise ValueError("weight must be 0..100")

        now = ts_ms if ts_ms is not None else int(time.time() * 1000)
        nonce = os.urandom(12)

        header: Dict[str, Any] = {
            "v": self.cfg.protocol_version,
            "alg": self.cfg.alg,
            "cid": self.cid,
            "kid": self.cfg.key_id,
            "ts": now,
            "nonce": _b64(nonce),
            "seq": self.seq.next(),
        }

        payload: Dict[str, Any] = {
            "shelf": shelf,
            "code": code,
            "weight": weight,
            "domain_hash": self.derive_domain_hash(domain_id),
            "doc_fingerprint": self.derive_doc_fingerprint(doc_vector_bytes),
            "event": event,
        }

        if ctx_window is not None:
            payload["ctx"] = {"window": ctx_window, "flags": int(ctx_flags)}

        sig = self.sign_packet(header, payload)
        return {"header": header, "payload": payload, "auth": {"sig": sig}}

def build_doc_vector_bytes(
    doc_type_id: int,
    issuer_role_id: int,
    impact_band_id: int,
    lifecycle_state_id: int,
    local_doc_id_salted: Optional[bytes] = None,
) -> bytes:
    parts = [
        doc_type_id.to_bytes(2, "big", signed=False),
        issuer_role_id.to_bytes(2, "big", signed=False),
        impact_band_id.to_bytes(1, "big", signed=False),
        lifecycle_state_id.to_bytes(1, "big", signed=False),
    ]
    if local_doc_id_salted:
        parts.append(local_doc_id_salted)
    return b"".join(parts)

def salt_local_id(csalt_b64: str, local_id: str) -> bytes:
    csalt = base64.b64decode(csalt_b64)
    return _sha256(b"WINDI:LOCALID:v1" + csalt + local_id.encode("utf-8"))

File: SDK_v1.1_RFC003/windi-sdk-v1/emitters/node/windi-emitter.ts
// WINDI SDK - Reference Emitter (TypeScript / Node)
// RFC-002: WINDI-SEC-JSON-v1 (HMAC default)
// "Flow is Truth. Content is Sovereign."

import crypto from "crypto";

export type Alg = "HMAC-SHA256";
export type Shelf = "S1" | "S2" | "S3" | "S4" | "S5" | "S6" | "S7";

export type WindiEvent =
  | "DOC_CREATED"
  | "APPROVAL_REQUESTED"
  | "APPROVED"
  | "REJECTED"
  | "APPROVAL_OVERRIDDEN"
  | "DEADLINE_EXCEEDED"
  | "DEPENDENCY_LINKED"
  | "DEPENDENCY_BLOCKING"
  | "STATE_TRANSITION";

const EVENTS = new Set<WindiEvent>([
  "DOC_CREATED",
  "APPROVAL_REQUESTED",
  "APPROVED",
  "REJECTED",
  "APPROVAL_OVERRIDDEN",
  "DEADLINE_EXCEEDED",
  "DEPENDENCY_LINKED",
  "DEPENDENCY_BLOCKING",
  "STATE_TRANSITION",
]);

export interface WindiEmitterConfig {
  clientId: string;
  keyId: string;
  csaltB64: string;
  hmacKeyB64: string;
  protocolVersion?: "1.0";
  alg?: Alg;
}

export interface TelemetryHeader {
  v: "1.0";
  alg: Alg;
  cid: string;
  kid: string;
  ts: number;
  nonce: string;
  seq: number;
}

export interface TelemetryPayload {
  shelf: Shelf;
  code: string;
  weight: number;
  domain_hash: string;
  doc_fingerprint: string;
  event: WindiEvent;
  ctx?: { window: "1m" | "5m" | "15m" | "1h" | "24h" | "7d" | "30d"; flags: number };
}

export interface TelemetryPacket {
  header: TelemetryHeader;
  payload: TelemetryPayload;
  auth: { sig: string };
}

function b64(buf: Buffer): string {
  return buf.toString("base64");
}

function sha256(data: Buffer | string): Buffer {
  return crypto.createHash("sha256").update(data).digest();
}

function stableSort(value: any): any {
  if (Array.isArray(value)) return value.map(stableSort);
  if (value && typeof value === "object") {
    const out: any = {};
    Object.keys(value)
      .sort()
      .forEach((k) => (out[k] = stableSort(value[k])));
    return out;
  }
  return value;
}

function canonicalJson(obj: any): Buffer {
  return Buffer.from(JSON.stringify(stableSort(obj)), "utf8");
}

export class WindiEmitter {
  private readonly cfg: Required<WindiEmitterConfig>;
  private readonly csalt: Buffer;
  private readonly hmacKey: Buffer;
  private readonly cid: string;
  private seq: number;

  constructor(cfg: WindiEmitterConfig) {
    this.cfg = { protocolVersion: "1.0", alg: "HMAC-SHA256", ...cfg };
    this.csalt = Buffer.from(this.cfg.csaltB64, "base64");
    this.hmacKey = Buffer.from(this.cfg.hmacKeyB64, "base64");
    this.cid = b64(sha256(Buffer.from(this.cfg.clientId, "utf8")));
    this.seq = 0;
  }

  deriveDomainHash(domainId: string): string {
    const raw = Buffer.concat([
      Buffer.from("WINDI:DOMAIN:v1", "utf8"),
      this.csalt,
      Buffer.from(domainId, "utf8"),
    ]);
    return b64(sha256(raw));
  }

  deriveDocFingerprint(docVectorBytes: Buffer): string {
    const raw = Buffer.concat([
      Buffer.from("WINDI:DOC:v1", "utf8"),
      this.csalt,
      docVectorBytes,
    ]);
    return b64(sha256(raw));
  }

  private sign(header: TelemetryHeader, payload: TelemetryPayload): string {
    const msg = canonicalJson({ header, payload });
    const mac = crypto.createHmac("sha256", this.hmacKey).update(msg).digest();
    return b64(mac);
  }

  emit(params: {
    shelf: Shelf;
    code: string;
    weight: number;
    domainId: string;
    docVectorBytes: Buffer;
    event: WindiEvent;
    ctxWindow?: TelemetryPayload["ctx"]["window"];
    ctxFlags?: number;
    tsMs?: number;
  }): TelemetryPacket {
    const { shelf, code, weight, domainId, docVectorBytes, event } = params;
    if (!EVENTS.has(event)) throw new Error(`Invalid event: ${event}`);
    if (weight < 0 || weight > 100) throw new Error("weight must be 0..100");

    const ts = params.tsMs ?? Date.now();
    const nonce = crypto.randomBytes(12);

    const header: TelemetryHeader = {
      v: "1.0",
      alg: "HMAC-SHA256",
      cid: this.cid,
      kid: this.cfg.keyId,
      ts,
      nonce: b64(nonce),
      seq: ++this.seq,
    };

    const payload: TelemetryPayload = {
      shelf,
      code,
      weight,
      domain_hash: this.deriveDomainHash(domainId),
      doc_fingerprint: this.deriveDocFingerprint(docVectorBytes),
      event,
      ctx: { window: params.ctxWindow ?? "5m", flags: params.ctxFlags ?? 0 },
    };

    const sig = this.sign(header, payload);
    return { header, payload, auth: { sig } };
  }
}

export function saltLocalId(csaltB64: string, localId: string): Buffer {
  const csalt = Buffer.from(csaltB64, "base64");
  return sha256(Buffer.concat([Buffer.from("WINDI:LOCALID:v1", "utf8"), csalt, Buffer.from(localId, "utf8")]));
}

export function buildDocVectorBytes(params: {
  docTypeId: number;
  issuerRoleId: number;
  impactBandId: number;
  lifecycleStateId: number;
  localDocIdSalted?: Buffer;
}): Buffer {
  const buf = Buffer.alloc(6);
  buf.writeUInt16BE(params.docTypeId, 0);
  buf.writeUInt16BE(params.issuerRoleId, 2);
  buf.writeUInt8(params.impactBandId, 4);
  buf.writeUInt8(params.lifecycleStateId, 5);
  return params.localDocIdSalted ? Buffer.concat([buf, params.localDocIdSalted]) : buf;
}
